### ISSUE TITLE
Files with same name break permissions page

### ISSUE THREAD

**Issue by qqmyers** [IQSS/dataverse#6401]
QDR has Datasets where a restricted file in one version of dataset has been deleted and then a new file with the same name added to a later version (e.g. with curatorial changes). The Edit/Permissions/File page only shows one file with this name and appears to show metadata that mixes the state of the two files. In our case, the old deleted file was accessible by :authenticateduser but the table shows Deleted and "0 Users/Groups", the latter of which is only true of the new file (which is not Deleted). Attempting to give :authenticateduser download permissions results in a database uniqueness error - the file id getting called is the old file and despite the table shown 0 groups, it is already accessible by :authenticateduser. (Consequently, there is no way to assign permissions on the new file at all.)

After some digging, I tracked this down to the use of a TreeMap with DataFile keys conflicting with the DataFile.comparesTo method which only compares displayNames and hence tells the TreeMap these files are equal. Since TreeMaps require unique keys, it only keeps one entry and it actually appears to swap the DataFile used as a key without updating the value (why not - they're equal!), resulting in the mixed up info.

I've currently fixed this for QDR by writing a 'strict' comparator, used only in this one TreeMap, that orders by display name but then orders by id if the displayNames are equal. This appears to fix the problem.

It may be that this issue is unique to the permissions page and only affects QDR due to the practice described above. However, since the DataFile.comparesTo method should be getting used in any classes trying to order the files, I'm not sure if this affects anything else. If files with the same name but, for example, with different paths, end up being allowed, this comparesTo method could be problematic elsewhere. Conversely, unless there are cases where we're relying on comparesTo to drop duplicate files, making it strict - using the id to break ties - might be the more general approach/help prevent future issues. (I doubt its good practice to make equals() and comparesTo() inconsistent about equality...)

I'll be making a PR. If that should be a change to DataFile.comparesTo() rather than a fix specifically for the permissions page, let me know.

**Comment by scolapasta** [comment]
Hi @qqmyers 
We actually have this before (https://github.com/IQSS/dataverse/issues/4287) and it's an issue we've wanted to prioritize but has gotten bumped a few times. So I'm glad you've been able to look into it and will have a PR. 

You are correct in that having equals() and comparesTo() inconsistent about equality is not a good practice and we should avoid. I would prefer we have a solution that covers the whole app and make sure that other places using the general compare to continue to work or are fixed accordingly.

**Comment by kcondon** [comment]
@qqmyers @scolapasta I was not able to reproduce the above problem. Can you provide steps to reproduce?

What I did:
1. upload fileA, make restricted, assign file level perms to :authenticateusers, publish
2. delete fileA, upload fileA again, assign no perms, publish. perms to fileA still appears in file perms table 
3. restrict fileA, publish, :authenticateduser perms still on fileA but when viewed in UI it is clear perms are on the first fileA from step 1, not the current one. Same behavior on your branch.

4. tried assigning file downloader perms to :authenticatedusers at the dataset level on develop and this branch and worked in both cases.

What am I missing?

**Comment by qqmyers** [comment]
@kcondon - I'm not seeing an obvious problem with your recipe. As soon as there are two files with the same name that are both restricted, the new branch should show both copies in the table. At QDR (with some style changes) I see 
![image](https://user-images.githubusercontent.com/6731983/69650255-53e7be00-103c-11ea-8a34-9773b9cd3e17.png) after the fix. Before the table only showed 5 entries (I deleted 4 files and added 3, two with the same names, so two that have conflicts that the PR fixes).
In detail, I think I added the 4 files, restricted and set perms, then published v1. deleted the files and published v2, and then I've been adding, restricting and looking at the permissions on the replacement files in the draft version. 
I guess it is still possible that there is another issue hiding multiple published versions, so I guess trying a new file with name conflict still in draft would be worth testing. Beyond that, I'm not sure what could be the difference. As usual, I can go back to the QDR code and make sure I haven't left something out in making the PR, but this is a fairly minimal change.
